coding/javascript.md

```diff
@@ -3,33 +3,93 @@ layout: base
 
 On all publicly accessible websites, core content and functionality should still be available without JavaScript. JavaScript applications written for a more specific audience which won't be indexed by search engines may be more lax, but should still treat progressive enhancement as a guiding principle.
 
-## Element hooks
+## Performance
+- Cache DOM queries — only select an element once.
+- Use event delegation as much as possible to reduce the number of events bound to the page
+- Understand when you cause an element repaint and reduce inline style manipulations accordingly.
+- If possible, always use CSS for element transforms and transitions.
 
+## Element hooks
 We should target page element in Javascript using the `data-js` element attribute instead of using the class attribute. Classnames should be used for styling only.
 
 e.g. `<a data-js="index-link" href="/index">Index</a>`.
 
-## Performance
+## Tooling
 
-- Cache DOM queries — only select an element once.
-- Use event delegation as much as possible to reduce the number of events bound to the page
-- Understand when you cause an element repaint and reduce inline style manipulations accordingly.
-- If possible, always use CSS for element transforms and transitions.
+### Dependency management
+We use [yarn](https://yarnpkg.com/en) for dependency management.
+
+### Testing
+We use the [jest](https://jestjs.io) unittest framework.
+
+### Code style & formatting
+To ensure code formatting consistency across our codebase, we use [Prettier](https://github.com/prettier/prettier).
+
+JS should also be written using [AirBnB](https://github.com/airbnb/javascript) style. New projects can adhere to this by selecting 'AirBnB' during `eslint init`, and existing projects can find the relevant eslint config in [eslint-config-airbnb](https://www.npmjs.com/package/eslint-config-airbnb).
+
+## React
+
+### Starting a new project
+When starting a new react application, we recommend [create-react-app](https://github.com/facebook/create-react-app) for bootstrapping the project. Avoid [ejecting](https://facebook.github.io/create-react-app/docs/available-scripts#npm-run-eject) as long as feasible to make updating dependencies easier.
+
+### Preferred libraries
+All projects should generally use the following:
+
+* style framework - [vanilla](https://github.com/vanilla-framework/vanilla-framework)
+* component testing - [enzyme](https://github.com/airbnb/enzyme)
+
+If you require routing, or state management:
+
+* routing - [react-router](https://github.com/ReactTraining/react-router)
+* app state management - [redux](https://redux.js.org)
+
+#### Redux specific
+* selector composition and memoisation - [reselect](https://github.com/reduxjs/reselect)
+* async/side-effects management (http, localstorage etc.) - [redux-saga](https://github.com/redux-saga/redux-saga)
+
+If you'd like introduce a new library, or feel we should replace one of the above, please make a PR to start a discussion.
```

Comment:
I think a lot of what redux offers can be achieved through the React context API and the upcoming hooks feature.

Reply:
> I think a lot of what Redux offers can be achieved through the React context API and the upcoming hooks feature.

Redux provides much more than just improving passing props through deeply nested components - time travel, state debugging via redux tools, and redux middleware (sagas, thunks) for decoupling code with side effects. Hooks are also currently an alpha feature.

Reply:
While Redux provides a lot of fancy features, its complexity is overkill for a vast majority of projects. These types of libraries should be chosen on a per-project basis and only after you've exhausted built-in options or simpler projects to backfill a few minor issues. 

With all that said....The line above `All projects should generally use the following:` should probably be removed.

Reply:
> The line above `All projects should generally use the following:` should probably be removed.

That line refers to vanilla an enzyme, which should be used in all react projects. I thought it was clear that the routing and state management were project dependent?

Reply:
I read it as the first statement under the heading for `Preferred Libraries` covering everything following. Probably my fault 😺